ISSUE-196-custom-file-storage-path --- provide options to set relative paths for storage of managed strawberryfield files and digital object files (json)

patdunlavey opened the pull request:
See #196 

This PR provides configuration options on the persistent file storage settings page to set relative paths for storage of the strawberryfield uploaded files (default none) and the digital object files (json) (default "dostorage").

patdunlavey commented:
@DiegoPino let me know your thoughts. Thanks!

DiegoPino commented on config/schema/strawberryfield.schema.yml:
```diff
@@ -79,9 +79,17 @@ strawberryfield.storage_settings:
     file_scheme:
       type: string
       label: 'Storage Scheme for Persisting Files'
+    file_path:
```
Ok, do we need both? if we add here the file_path then the alter hook needs to be adapted again? Just wondering since the ISSUE was about the Object File Path. But happy to deal with both (we need an alter too in the form)

patdunlavey replied:
Yes, this is not strictly necessary since we can modify the path through hook_strawberryfield_file_destination_alter(). However, it seems to me that being able to set the path would be desired pretty often, and requiring a user to implement a hook just to set this to a different path seems a bit onerous, particularly as they will see that we do give them the configuration option for the digital object file storage.
I don't feel strongly about this though, Happy to take it out if you think that makes more sense.

DiegoPino replied:
Ok, let's keep it. I understand how this can be useful for vanilla implementations!

DiegoPino commented on src/EventSubscriber/StrawberryfieldEventInsertSubscriberDepositDO.php:
```diff
@@ -91,6 +98,9 @@ public function __construct(
     )->get('object_file_scheme');
+    $this->destinationRelativePath = $config_factory->get(
+      'strawberryfield.storage_settings'
+    )->get('object_file_path') ?? "dostorage";
```
This looks good. What if you make `dostorage` a class constant?

patdunlavey replied:
Done.

DiegoPino commented on src/Form/FilePersisterServiceSettingsForm.php:
```diff
@@ -47,7 +47,12 @@ public function buildForm(array $form, FormStateInterface $form_state) {
+    $form['file_path'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Relative Path for Persisting Files'),
+      '#description' => $this->t('Path relative to the root of the storage scheme selected above where the hashed directories used to store persisted managed files will be created. Do not include beginning or ending slashes. Default is "".'),
```
Let's add a warning here. That says "changing this path after your ingests won't modify existing  Files"

patdunlavey replied:
Ok, I added this note to both fields. Tell me if that's not correct!

DiegoPino replied:
Thanks. Correct!

DiegoPino commented on src/Form/FilePersisterServiceSettingsForm.php:
```diff
@@ -47,7 +47,12 @@ public function buildForm(array $form, FormStateInterface $form_state) {
+      '#type' => 'textfield',
+      '#title' => $this->t('Relative Path for Persisting Files'),
+      '#description' => $this->t('Path relative to the root of the storage scheme selected above where the hashed directories used to store persisted managed files will be created. Do not include beginning or ending slashes. Default is "".'),
+      '#default_value' => !empty($config_storage->get('file_path')) ? $config_storage->get('file_path'): "",
```
Wonder if we should better go for NULL instead of empties? And then change the logic of the setting? Not sure really

patdunlavey replied:
My thinking was that the user input for this field is trimmed, exploded on slashes, etc - all operations that expect a string.

DiegoPino replied:
Ok, let's roll with an empty string

DiegoPino commented on src/Form/FilePersisterServiceSettingsForm.php:
```diff
@@ -306,9 +317,11 @@ public function submitForm(array &$form, FormStateInterface $form_state) {
       ->save();
     $this->config('strawberryfield.storage_settings')
       ->set('file_scheme', $form_state->getValue('file_scheme'))
+      ->set('file_path', trim($form_state->getValue('file_path')," \n\r\t\v\0/"))
```
Lovely! But I think for this case we need to make sure these paths are sanitized. Also S3 has size limits. Look at the method where I refactored the hook. I have a few notes there about this

patdunlavey replied:
So checkPlain and limit the path to 128 characters? Ok to just set the field character limit?

DiegoPino replied:
I think we need to make folder validation method. Should not have spaces. Should have a controlled list of characters (like no utf8 strawberries).  Here is a list of what needs to be checked. I agree we are also allowing local storage, e.g what @giancarlobi does, but still not sure I want to have on a local unix filesystem something that fights a 1986 unix spec too much. See https://www.ezs3.com/public/232.cfm

patdunlavey replied:
Ok, I'll work on a validation method.

DiegoPino commented on src/StrawberryfieldFilePersisterService.php:
```diff
@@ -223,7 +231,10 @@ public function getDestinationUri(
+    // First, get any parent directories.
+    $relativefolders = explode('/', $this->storageConfig->get('file_path') ?? "");
+    $relativefolders[] = substr($checksum, 0, 3);
+    $relativefolder = implode('/', array_filter($relativefolders));
```
Ok, this needs to be added to the refactor too

patdunlavey replied:
I'm thinking let's get your refactor merged first, and then resolve this branch it. Does that make sense?

DiegoPino replied:
Ok, let's roll with that. Will wait for your review on that pull before merging it. Thx

DiegoPino commented:
@patdunlavey thanks for the pull. I added a few questions really more than code reviews. Looks good, i would mostly like to do some more sanitization on the inputs and make sure there is no way a user can corrupt storage via this forms. This also means I need to refactor my #199 to match the new paths.

patdunlavey commented:
Update: #199 has been merged. I need to address the file path sanitation issues (in the next day or so) and then will ask for further review.

patdunlavey commented:
I've added validation for the two file path fields - both ajax and upon submit. It's probably overkill, but it's working pretty good for me. The one thing that had me scratching my head for a good while was how to validate the system file path (public and private schemes). As I commented in the code, neither \Drupal::service('path.validator')->isValid() nor \Drupal::service('stream_wrapper_manager')->isValidUri() work --- at all! Well, I probably don't understand how they're supposed to work. I ended up using FileSystem::prepareDirectory() to test if the directory path provided could be written to. Can you review that part in particular, @DiegoPino ? There is almost certainly a better way to do it!

patdunlavey commented:
Hi @DiegoPino. With all the discussion of the CSV exporter today I don't want to lose sight of this issue being the only one we know of that is actually blocking us from starting to ingest content into our production instance. Any chance you could take a bit of time to give this a look?

DiegoPino commented:
@patdunlavey I can tomorrow. Did you adapt the code so it merges clean with the last hook change? This will break my open https://github.com/esmero/strawberryfield/pull/189 so give me at least 24 hours, hands totally full. Will check back when ready. thx

patdunlavey commented:
@DiegoPino yes it merges clean with the last hook change (that change has been merged to 1.0.0-RC3 and this PR is against that). I'll take a look at #189. Maybe I can create a combined merge branch? Or if that branch is ready to merge, maybe you could do that and then I can reconcile again with this branch?

DiegoPino commented on src/Form/FilePersisterServiceSettingsForm.php:
```diff
@@ -56,6 +73,22 @@ public function buildForm(array $form, FormStateInterface $form_state) {
+      '#ajax' => [
+        'callback' => [$this, 'validateObjectFilePath'],
+        'effect' => 'fade',
+        'wrapper' => 'objectfilepath-exec-path-validation',
```
The `exec` is a bit confusing. Better remove since there is no execution at all

patdunlavey replied:
Yep, a lazy copy-paste mistake.

patdunlavey replied:
Fixed.

DiegoPino commented on src/EventSubscriber/StrawberryfieldEventInsertSubscriberDepositDO.php:
```diff
@@ -91,6 +103,9 @@ public function __construct(
     $this->destinationScheme = $config_factory->get(
       'strawberryfield.storage_settings'
     )->get('object_file_scheme');
+    $this->destinationRelativePath = $config_factory->get(
```
Like this

DiegoPino commented on src/EventSubscriber/StrawberryfieldEventInsertSubscriberDepositDO.php:
```diff
@@ -20,6 +20,11 @@ class StrawberryfieldEventInsertSubscriberDepositDO extends StrawberryfieldEvent
+  /**
+   * Default path for digital object file storage.
+   */
+  public const DEFAULTOBJECTSTORAGEFILEPATH = 'dostorage';
```
really like this. Could you add a few "_" to make it more legible? Sorry. It really helps

patdunlavey replied:
Better now, I hope!

DiegoPino commented on config/schema/strawberryfield.schema.yml:
```diff
@@ -79,9 +79,17 @@ strawberryfield.storage_settings:
+    object_file_path:
+      type: string
+      label: 'Storage Path under the Storage Scheme, for Persisting Digital Objects in JSON format'
+      default: "dostorage"
```
good!

DiegoPino commented on src/StrawberryfieldUtilityService.php:
```diff
@@ -28,6 +28,20 @@ class StrawberryfieldUtilityService implements StrawberryfieldUtilityServiceInte
+   * File system path pattern. Must begin and end with a word character, may have
+   * forward slashes delimiting directory names within the path.
+   */
+  const PATHPATTERN = "/(^(\w)+$)|(^(\w)([\/\w-]*)(\w)$)/";
```
This is nice. Could we make this one more explicit? VALID_FILESYSTEM_PATH_PATTERN?

patdunlavey replied:
Better now, I hope.

DiegoPino commented on src/StrawberryfieldUtilityService.php:
```diff
@@ -28,6 +28,20 @@ class StrawberryfieldUtilityService implements StrawberryfieldUtilityServiceInte
+   * See \Drupal\strawberryfield\StrawberryfieldUtilityServiceInterface::s3FilePathIsValid
+   * for explanation of the rules.
+   */
+  const S3PATHPATTERN = "/^[a-z0-9][a-z0-9\-]{1,61}[a-z0-9]$/";
```
Same here (also super nice)  VALID_S3_PREFIX_PATTERN? (since in S3 paths are normally known as prefixes)

patdunlavey replied:
Better now I hope.

DiegoPino reviewed (changes requested):
@patdunlavey this is pretty much in good shape but still needs a bit more work. There is a bit of redundancy in the code (validation methods can be a single) one and S3 v/s any other are handled differently when it does not need to. Some CONST (thanks for that) need spacing (_) to be readable and a few minor things. I can help you tomorrow to get those done and we can merge. Thanks again for working on this.

patdunlavey commented on src/StrawberryfieldUtilityService.php:
```diff
@@ -329,4 +343,57 @@ public function getCountByProcessorInSolr(EntityInterface $entity, string $proce
+      $valid = \Drupal::service('file_system')->prepareDirectory($test_path, FileSystemInterface::CREATE_DIRECTORY);
+      if($valid) {
+        // Successfully created the directory, so we delete it now.
+        \Drupal::service('file_system')->rmdir($test_path);
```
Thanks for the dependency injection correction - this symfony stuff is still very foggy to me!
I'm with you that creating a directory and then deleting it seems dangerous. I was hoping you could point me to another way to determine if the proposed destination path is writable!
I'll leave the two-step logic as-is unless and until we can think of something more elegant.

patdunlavey replied:
Changes committed. Please re-review.

patdunlavey commented on src/Form/FilePersisterServiceSettingsForm.php:
```diff
@@ -145,10 +178,155 @@ public function buildForm(array $form, FormStateInterface $form_state) {
+    }
+    else {
+      switch ($scheme) {
+        case 's3':
```
I'm nervous about removing the regex tests because I'm not 100% sure I like the invasive method of determining writability. Am I just being silly?

DiegoPino replied:
Oh, do not remove them. What I say here is that you can deal with any wrapper in the same way. and to be honest, to allow portability between filesystem, the MOST restrictive regex should be the one that is applied to both (so the S3). That makes also your code simpler @patdunlavey . That way moving from local filesystem to the cloud is seamless. If not you have to update all your ADOS!

patdunlavey replied:
Ok, switched to using the S3 prefix pattern for all. Please re-review!

patdunlavey commented on src/Form/FilePersisterServiceSettingsForm.php:
```diff
@@ -145,10 +178,155 @@ public function buildForm(array $form, FormStateInterface $form_state) {
+      }
+    }
+    if (!$valid) {
+      $warning_message = self::filePathErrorMessage('object file', $scheme,
```
Fixed.

patdunlavey commented on src/Form/FilePersisterServiceSettingsForm.php:
```diff
@@ -145,10 +178,155 @@ public function buildForm(array $form, FormStateInterface $form_state) {
+      }
+    }
+    if (!$valid) {
+      $warning_message = self::filePathErrorMessage('file', $scheme, $path);
```
Fixed.

patdunlavey commented on src/Form/FilePersisterServiceSettingsForm.php:
```diff
@@ -145,10 +178,155 @@ public function buildForm(array $form, FormStateInterface $form_state) {
+   *
+   * @return mixed
+   */
+  private function filePathErrorMessage($type, $scheme, $path) {
```
See latest commit. Fairly major refactoring of a lot of this stuff since we are changing to using the same validation rule for S3 and local file systems.

patdunlavey commented:
I think I have addressed all your review comments. The main thing was refactoring to only use one validation method (S3). Got rid of a fair bit of redundant code. Made the error message even more wordy! Ready for your re-review when you get a chance. Thanks so much @DiegoPino , you are amazing!

DiegoPino commented:
@patdunlavey amazing! Will give it a spin at 2:30 and will let you know. Thanks!!

DiegoPino commented on src/Form/FilePersisterServiceSettingsForm.php:
```diff
@@ -471,15 +426,14 @@ public function validateForm(array &$form, FormStateInterface $form_state) {
-            $valid = \Drupal::service('strawberryfield.utility')
-              ->internalFilePathIsValid($scheme, $path);
-            break;
+        // Validate for known schemes.
```
@patdunlavey this looks much better. I will leave this at it is but will refactor tonight a bit. Why?

We have this here:
 $scheme_options = OcflHelper::getVisibleStreamWrappers();

If a scheme is not in that list, then it is actually invalid and we need to validate as "false". No worries. It works the way you wrote but I feel its good we reuse code.

patdunlavey replied:
Makes sense. Thanks!

DiegoPino commented on src/StrawberryfieldUtilityService.php:
```diff
@@ -97,15 +93,22 @@ class StrawberryfieldUtilityService implements StrawberryfieldUtilityServiceInte
    *   The Search API parse Manager
    */
-  public function __construct(FileSystemInterface $file_system, EntityTypeManagerInterface $entity_type_manager, ConfigFactoryInterface $config_factory, ModuleHandlerInterface $module_handler, EntityFieldManagerInterface $entity_field_manager, ParseModePluginManager $parse_mode_manager) {
+  public function __construct(
```
@patdunlavey I see a lot of DCS changes here. Some super good (thx) other a bit crazy and overly strict. I will also give this file a second pass after the merge to restore some of lost read-ability (but still keeping it DCS)

patdunlavey replied:
Sorry about that. I didn't notice until I had already committed and pushed. If you want, I can reverse these.

DiegoPino replied:
No worries. Php Storm does a lot to conform to the 80 lines limit but DCS does not apply that rule for functions. I can do that

DiegoPino commented on src/StrawberryfieldUtilityService.php:
```diff
@@ -163,25 +167,28 @@ public function getStrawberryfieldMachineNames() {
       ->condition('entity_type', 'node')
       ->condition('bundle', $bundle)
-      ->condition('field_type' , 'strawberryfield_field')
+      ->condition('field_type', 'strawberryfield_field')
```
nice! gosh.

DiegoPino commented on src/StrawberryfieldUtilityService.php:
```diff
@@ -191,24 +198,29 @@ public function getStrawberryfieldConfigFromStorage($bundle = 'digital_object')
    * {@inheritdoc}
    */
-  public function getStrawberryfieldMachineForBundle($bundle = 'digital_object') {
+  public function getStrawberryfieldMachineForBundle($bundle = 'digital_object'
```
@patdunlavey like this one. Totally crazy how PHP Storm reformatted this

DiegoPino commented on src/StrawberryfieldUtilityService.php:
```diff
@@ -246,7 +260,7 @@ public function getStrawberryfieldSolrFields(Index $index_entity) {
    * {@inheritdoc}
    */
-  public function verifyCommand($execpath) :bool {
+  public function verifyCommand($execpath): bool {
```
good!

DiegoPino commented on src/StrawberryfieldUtilityService.php:
```diff
@@ -346,54 +367,48 @@ public function getCountByProcessorInSolr(EntityInterface $entity, string $proce
+  public function filePathIsValid(string $scheme, string $path): bool {
     // Empty path is always valid.
-    if(empty($path)) {
+    if (empty($path)) {
```
@patdunlavey You trimming this somewhere right? Like "       " becomes "" ?

DiegoPino commented on src/StrawberryfieldUtilityService.php:
```diff
@@ -346,54 +367,48 @@ public function getCountByProcessorInSolr(EntityInterface $entity, string $proce
-  public function s3FilePathIsValid(string $path): bool {
-    // Empty path is always valid.
-    if(empty($path)) {
+    $known_schemes = array_merge(['s3'], array_keys(\Drupal::service('stream_wrapper_manager')->getWrappers(StreamWrapperInterface::LOCAL)));
```
Again, we have the OCFL toolbox I wrote to fetch all known schemes that are active and can be used to writing. Will refactor

patdunlavey replied:
Ok, thanks for your patience with a D8+ noob!

DiegoPino commented on src/StrawberryfieldUtilityService.php:
```diff
@@ -346,54 +367,48 @@ public function getCountByProcessorInSolr(EntityInterface $entity, string $proce
+        $this->fileSystem->rmdir($test_path);
+      }
+    }
+    return !empty($valid);
```
@patdunlavey better (for readability) to set $valid to FALSE at the beginning. That way you always return$valid

DiegoPino commented:
@patdunlavey so I added a few comments. I will merge at it is tonight but will do a small refactor. The DCS pass on the strawberryfield utilities is a bit crazy and there are a few pieces of code that can be simplified by reusing (see same form) existing methods (visible stream wrappers). Thanks. Please do not merge until I refactor after merging. Will mention you. Hugs!

patdunlavey commented:
Hugs backatcha!

DiegoPino reviewed (approved):
ready to merge. Small refactor coming later to night! Thanks @patdunlavey on your work on this!
